Implemented tf.SparseTensor.from_dense(np.ndarray) function that returns a SparseTensor from a Dense numpy array

joglekara opened the pull request:
Implemented a tf.SparseTensor.from_dense() function
this PR adds two things.

- It adds a function that takes in a numpy.ndarray() and returns a tf.SparseTensor.
It uses scipy's coo_matrix capabilities.
- It adds a test for the above.

Commit:
Implemented a tf.SparseTensor.from_dense() function

this commit adds two things.

- It adds a function that takes in a numpy.ndarray() and returns a tf.SparseTensor.
It uses scipy's coo_matrix capabilities.
- It adds a tset for the above.

Commit:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into fromdense

jsimsa commented:
@rthadur I am not familiar with this part of the code base but perhaps @edloper could take a look.

rthadur commented:
@jsimsa thank you !

mellanox-github commented:
Can one of the admins verify this patch?

edloper commented:
In general, I would expect TensorFlow functions to take Tensor inputs.  (Numpy arrays should also be accepted as inputs, but should be converted to Tensors.)  This allows them to be used in TensorFlow graphs that are run in graph mode.

In particular, I would expect a method "tf.sparse.from_dense(x)" to accept a dense Tensor, and return a SparseTensor.  This function is already implemented as [tfp.math.dense_to_sparse](https://www.tensorflow.org/probability/api_docs/python/tfp/math/dense_to_sparse).  

It's possible that dense_to_sparse should exist in the main tensorflow package (as opposed to tfp).  I don't have much of an opinion on that one way or the other.  If you'd like to advocate for having dense_to_sparse in the main tensorflow package, I'll check with the TF api owners to see what they think.

joglekara commented:
Hmm. Thanks for bringing up the tfp function, I was not aware of that implementation.

So yes, I agree with your final comment, we just have to decide whether this should be included in the main tensorflow package. 

It makes sense to me to include it, mostly because tfp doesn't really have the same capability for sparse math that the main tensorflow package does. In fact, it only has 2 sparse-related operations, one of which is in the main tf package already (matmul). 

The way it would work now would be to use tfp for just this one application, and then use tf for the rest of your work.

Given the above, it makes sense to include this functionality in the main package

joglekara commented:
@edloper Any further thoughts on this?

jsimsa commented:
@joglekara I will bring this to the attention of TF API owners and get back to you.

karmel commented:
(for API owners) Should this be SparseTensor.from_tensor() to match RaggedTensor + other composite tensor types?

Implementation-wise, note that this should not rely on scipy, and should support tensors as well.

Commit:
removed scipy dependency and added method for creating SparseTensor from a tensor

The SciPy dependency was removed by writing a function that extracts the shape, row, col indices, and non-zero values of a dense numpy array.

the method for creating a SparseTensor from a tensor is borrowed from tf 1.15 release, from the function that also happens to be named from_dense().

joglekara commented:
Thanks for the comment @karmel .

I've made the requested modifications. Conveniently enough, some of this functionality has been implemented in the future versions of tensorflow, but this PR contributes the ability to convert from a numpy array directly to a SparseTensor.

Commit:
Update sparse_tensor_test.py

Updated with a test that doesn't include scipy, and even more straightforward.

joglekara commented:
@rthadur @jsimsa  @karmel Any further thoughts on this?

alextp commented on tensorflow/python/framework/sparse_tensor.py:
```diff
@@ -417,3 +420,62 @@ def is_sparse(x):
+def from_dense(dense, name=None):
+    """Convert a dense `np.array` or a dense tensor to a `tf.SparseTensor`
+
+    The code to convert a dense tensor to a `SparseTensor` is borrowed from
```
This should not be in the docstring

joglekara replied:
Hmm, it is not clear to me what should not be included in the docstring. Can you please elaborate?

alextp replied:
I meant the paragraph about borrowed code. Users don't care.

alextp commented on tensorflow/python/framework/sparse_tensor.py:
```diff
@@ -417,3 +420,62 @@ def is_sparse(x):
   return isinstance(x, (SparseTensor, SparseTensorValue))
+
+def from_dense(dense, name=None):
+    """Convert a dense `np.array` or a dense tensor to a `tf.SparseTensor`
```
Why can't we use the from_tensor to convert from a numpy array (since we already have conversions from numpy array to dense tensor with zero copies)?

joglekara replied:
My understanding is that making a dense tensor first, and then using from_tensor, results in allocating a dense tensor on the GPU. Often, we'd want to avoid making a dense tensor at all because the tensor size might be too large to comfortably work on a GPU. 

Therefore, we want to skip the dense tensor step completely, and initialize a SparseTensor as the first instantiation of the data on a GPU. Does that make sense?

alextp replied:
You can allocate the dense tensor on the CPU though if you do tf.device(cpu): from_tensor

joglekara replied:
I just tested this on Colab and I believe you're correct!

Thanks for reminding me of this functionality. I'm going to close the PR.

joglekara commented:
Closing PR because reviewer alerted me to a simple enough workaround.
